dotCMS/src/main/java/com/dotcms/rest/api/v1/portlet/PortletResource.java

```diff
@@ -0,0 +1,154 @@
+      APILocator.getLanguageAPI();
+  
+  /**
+   * Default class constructor.
+   */
+  public PortletResource() {
+    this(new WebResource(new ApiProvider()), APILocator.getPortletAPI());
+  }
+
+  @VisibleForTesting
+  public PortletResource(WebResource webResource, PortletAPI portletApi) {
+    this.webResource = webResource;
+    this.portletApi = portletApi;
+  }
+
+  @POST
+  @Path("/custom")
+  @JSONP
+  @NoCache
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+  public final Response contentPortlet(@Context final HttpServletRequest request,
+                                 final CustomPortletForm formData) {
+    InitDataObject init = webResource.init(null, true, request, true, "roles");
+
+    Response response = null;
+
+    try {
+
+      if (!UtilMethods.isSet(formData.portletId)) {
+        return ExceptionMapperUtil.createResponse(new DotStateException("Portlet Id is required"),
+                Response.Status.BAD_REQUEST);
+      }
+      if (!UtilMethods.isSet(formData.portletName)) {
+        return ExceptionMapperUtil.createResponse(new DotStateException("Portlet Name is required"),
+                Response.Status.BAD_REQUEST);
+      }
+
+      final String portletId = "c-" + formData.portletId;
+      Portlet portlet = portletApi.findPortlet(portletId);
+      if (portlet != null) {
+        return ExceptionMapperUtil.createResponse(new DotStateException("PortletId already Exists"),
+                Response.Status.CONFLICT);
+      }
+      List<String> contentTypes = formData.resolveContentTypes().stream().map(ct -> ct.variable())
+              .collect(Collectors.toList());
+      List<String> baseTypes = formData.resolveBaseTypes().stream().map(bt -> bt.name())
+              .collect(Collectors.toList());
+
+      if (contentTypes.size() + baseTypes.size() == 0) {
+        return ExceptionMapperUtil.createResponse(
+                new DotStateException("You must specify at least one baseType or Content Type"),
+                Response.Status.BAD_REQUEST);
+      }
+
+      final Portlet contentPortlet = portletApi.findPortlet("content");
+      Map<String, String> initValues = new HashMap<>();
+
+      initValues.putAll(contentPortlet.getInitParams());
+      initValues.put("name", formData.portletName);
+      initValues.put("baseTypes", String.join(",", baseTypes));
+      initValues.put("contentTypes", String.join(",", contentTypes));
+      initValues.put("portletSource", "db");
+
+      Portlet newPortlet = new DotPortlet(portletId, contentPortlet.getPortletClass(), initValues);
+
+      APILocator.getPortletAPI().savePortlet(newPortlet);
+
+      Map<String, String> keys = ImmutableMap
+              .of(com.dotcms.repackage.javax.portlet.Portlet.class.getPackage().getName()
+                      + ".title." + portletId, formData.portletName);
+      try {
+        for (Language lang : langApi.getLanguages()) {
```

Comment:
set to final and rename to language
in addition I think the whole for should be a transaction, you can use the LocalTransaction to wrap it